parse_cmd_test.go

```diff
@@ -1,151 +1,134 @@
-	for _, tt := range tableWithArgs {
-		t.Run(tt.testName, func(t *testing.T) {
-			gotCmd, gotArgs, gotErr := parseCmd(tt.inputStr)
-			if gotCmd != tt.wantedCmd || len(gotArgs) != len(tt.wantedArgs) {
-				t.Errorf("got %v, %v, wanted %v, %v\n", gotCmd, gotArgs, tt.wantedCmd, tt.wantedArgs)
+func Test_parseCmd(t *testing.T) {
+	for input, expected := range acceptedCommands {
+		t.Run(input, func(t *testing.T) {
+			cmd, args, err := parseCmd(input)
+			if err != nil {
+				t.Fatalf("unexpected error: %#+v", err)
 			}
 
-			switch gotCmd {
-			case "schedule":
-				for i, arg := range gotArgs {
-					if _, ok := daysList[arg]; !ok {
-						t.Errorf("Wrong argument %v for command %v\n", gotArgs[i], gotCmd)
-					}
-				}
-			case "skip":
-				if gotArgs[0] != "tomorrow" {
-					t.Errorf("Wrong argument %v for command %v\n", gotArgs[0], gotCmd)
-				}
-			case "unskip":
-				if gotArgs[0] != "tomorrow" {
-					t.Errorf("Wrong argument %v for command %v\n", gotArgs[0], gotCmd)
-				}
-			case "version":
-				// Always fine as long as the wantedCmd check passes.
-			default:
-				if gotCmd != "help" {
-					t.Errorf("unknown command %v\n", gotCmd)
-				}
-			}
+			assertEqual(t, cmd, expected.Cmd)
+			assertEqual(t, args, expected.Args)
+		})
+	}
 
-			_, ok := gotErr.(*parsingErr)
+	for _, input := range rejectedCommands {
+		t.Run(input, func(t *testing.T) {
+			cmd, args, err := parseCmd(input)
 
-			if tt.expectErr && !ok {
-				t.Errorf("Expected parsingErr but didn't get one\n")
-			} else if !tt.expectErr && ok {
-				t.Errorf("Got unexpected parsingError\n")
-			}
+			_, _ = assertErrorAs[*parsingErr](t, err)
+
+			assertEqual(t, cmd, "help")
+			assertEqual(t, args, nil)
 		})
 	}
 }
 
-var tableMisc = []struct {
-	testName   string
-	inputStr   string
-	wantedCmd  string
-	wantedArgs []string
-	expectErr  bool
-}{
-	{"command_is_superstring", "scheduleing monday", "help", nil, true},
-	{"command_is_substring", "schedul monday", "help", nil, true},
-	{"command_is_undefined", "mooh", "help", nil, true},
-	{"command_is_capitalized", "Help", "help", nil, false},
-}
+func assertEqual[T any](t *testing.T, a, b T) {
+	t.Helper()
 
-func TestParseCmdMisc(t *testing.T) {
-	for _, tt := range tableMisc {
-		t.Run(tt.testName, func(t *testing.T) {
-			gotCmd, gotArgs, gotErr := parseCmd(tt.inputStr)
-			if gotCmd != tt.wantedCmd {
-				t.Errorf("got %v, %v, wanted %v, %v\n", gotCmd, gotArgs, tt.wantedCmd, tt.wantedArgs)
-			}
-			_, ok := gotErr.(*parsingErr)
+	if reflect.DeepEqual(a, b) {
+		return
+	}
 
-			if tt.expectErr && !ok {
-				t.Errorf("Expected parsingErr but didn't get one\n")
-			} else if !tt.expectErr && ok {
-				t.Errorf("Got unexpected parsingError\n")
-			}
-		})
+	t.Errorf("expected %#+v to equal %#+v", a, b)
+}
+
+func assertErrorAs[T error](t *testing.T, err error) (target T, ok bool) {
+	ok = errors.As(err, &target)
+	if !ok {
+		t.Errorf("expected error as %T, got %#+v", target, err)
```

Comment:
I did a bit of a double-take on this, but it's just nits:

- "want" is a type, "got" is a value. It looks like `%T` for `err` would also give the concrete type ([playground](https://go.dev/play/p/2T-cKXnPoAL)) -- but I think an appropriate `%#v` will always ultimately indicate the type as well as the value, so yeah, this is fine.

- The [Go style guide](https://go.dev/wiki/TestComments#got-before-want) recommends "got before want", i.e. `got: %#+v of type %T, wanted type %T`. The direction is arbitrary (and these two `assert`s are internally consistent) -- I don't really care as long as we're consistent across the codebase, just wanted to note that the language's direction is the other way.

  (The style guide also has "got" and "want" instead of "expected" and "actual" -- which I also don't really care about.)

Reply:
Ah, this is more muscle-memory from testify's assertion helpers: [assert.Equal(t, expected, actual)](https://pkg.go.dev/github.com/stretchr/testify/assert#Equalf)

We can also blame long-term muscle memory coming from having written lots of RSpec tests (`expect(actual).to be(expected)`), if we want :smile: 

> I don't really care as long as we're consistent across the codebase, just wanted to note that the language's direction is the other way.

It seems I've _already_ introduced inconsistencies in how it seems I want to _use_ this helper (Zulip client tests call it both ways), so I'm going to have to fix something either way. So I might as well choose to match the language convention. I'll do this when I rewrite tests to use `cmp`.